### ISSUE TITLE
[autograd-custom-function] output variables' number check in backward path

### ISSUE THREAD

**Issue by arthuryuan1987** [pytorch/pytorch#60726]
## 🐛 Issue

I enable a custom operator LSTM on autograd dispatch key. I implement forward & backward function by following https://pytorch.org/tutorials/advanced/cpp_autograd.html#using-custom-autograd-function-in-c. In LSTM backward path, gradient weights are a set of variables, whose number is decided by layer number and bi-direction number of LSTM instance. The number is variable at runtime. I get an error at `torch/csrc/autograd/custom_function.h`,
```
    // NOLINTNEXTLINE(clang-diagnostic-sign-compare)
    if (num_outputs != num_forward_inputs) {
      std::string msg("function ");
      msg += name() + " returned an incorrect number of gradients (expected ";
      msg += c10::to_string(num_forward_inputs) + ", got " ;
      msg += c10::to_string(num_outputs) + ")";
      throw std::runtime_error(msg);
    }
```
That is because in my LSTM backward, I have more outputs number (split weight tensor list into tensors) than inputs number. But PyTorch inherent autograd function works. For example, `miopen_rnn_backward` works. In `variable_list MiopenRnnBackward::apply(variable_list&& grads)`,
```
    if (should_compute_output({ weight_ix })) {
      copy_range(grad_inputs, weight_ix, std::get<3>(grad_result));
    }
```
Gradient weights are split from the TensorList either, and put into variabl_list. So the final outputs' number of MiopenRnnBackward will be more than inputs'. The prototype is,

```
std::tuple<Tensor, Tensor, Tensor, std::vector<Tensor>> miopen_rnn_backward(
    const Tensor& input, TensorList weight, int64_t weight_stride0, const Tensor& weight_buf, const Tensor& hx, const c10::optional<Tensor>& cx_opt,
    const Tensor& output, const c10::optional<Tensor>& grad_output_r_opt, const c10::optional<Tensor>& grad_hy_r_opt, const c10::optional<Tensor>& grad_cy_r_opt, int64_t mode, int64_t hidden_size, int64_t num_layers, bool batch_first,
    double dropout, bool train, bool bidirectional, IntArrayRef batch_sizes, const c10::optional<Tensor>& dropout_state_opt,
    const Tensor& reserve, std::array<bool, 4> output_mask);
```
`weight` is TensorList in input parameter list, and flatten in output parameter list.

## Expected behavior

Can we remove the check of outputs' number in `autograd::custom_function`? 

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by jgong5** [comment]
@chunyuan-w

**Comment by albanD** [comment]
> Can we remove the check of outputs' number in autograd::custom_function?

No, that would not make sense. The custom Function should return gradients corresponding to the inputs of the forward.
Note that only the Tensors in the input arguments are considered. If you pass a vector<Tensor> to the input, the Tensors in the vector will be ignored (just like inputs in a list are ignored on the python side).

You will need to unpack that vector if you want to pass it as input to your custom function.

**Comment by arthuryuan1987** [comment]
But PyTorch in-source ATen operator `miopen_rnn` works?
In forward, `VariableType::miopen_rnn`, the `weight` is not unpacked, and is a TensorList.
```
std::tuple<at::Tensor,at::Tensor,at::Tensor,at::Tensor,at::Tensor> miopen_rnn(c10::DispatchKeySet ks, const at::Tensor & input, at::TensorList weight, int64_t weight_stride0, const at::Tensor & hx, const c10::optional<at::Tensor> & cx, int64_t mode, int64_t hidden_size, int64_t num_layers, bool batch_first, double dropout, bool train, bool bidirectional, at::IntArrayRef batch_sizes, const c10::optional<at::Tensor> & dropout_state) {
  // ...
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::redispatch::miopen_rnn(ks & c10::after_autograd_keyset, input_, weight_, weight_stride0, hx_, cx, mode, hidden_size, num_layers, batch_first, dropout, train, bidirectional, batch_sizes, dropout_state);
  })();
  // ...
}
```

And in its backward, `weight` is unpacked. So the number should be not same? But it works.
```
variable_list MiopenRnnBackward::apply(variable_list&& grads) {
  // ...
  auto grad_result = miopen_rnn_backward(input, weight, weight_stride0, result4, hx, cx, result0, grads[0], grads[1], grads[2], mode, hidden_size, num_layers, batch_first, dropout, train, bidirectional, batch_sizes, dropout_state, retain_variables ? result3.clone() : result3, grad_input_mask);
  // ...
  if (should_compute_output({ weight_ix })) {
    copy_range(grad_inputs, weight_ix, std::get<3>(grad_result));
  }
  return grad_inputs;
```

**Comment by arthuryuan1987** [comment]
PyTorch only counts the number of tensors in forward path as the inputs' number. We need also take tensors in TensorList,
```
struct ExtractVariables : IterArgs<ExtractVariables> {
  std::vector<bool>& is_var_;
  variable_list& list_;
  ExtractVariables(std::vector<bool>& is_var, variable_list& list) : is_var_(is_var), list_(list) {}
  void operator()(const c10::optional<at::Tensor>& x) {
    if (x.has_value() && x.value().defined()) {
      is_var_.push_back(true);
      list_.emplace_back(x.value());
    } else {
      is_var_.push_back(false);
    }
  }
  void operator()(const at::Tensor& x) {
    is_var_.push_back(true);
    list_.emplace_back(x);
  }
  template <typename T>
  void operator()(const T& x) {
    is_var_.push_back(false);
  }
};
```

```
template<class T>
template<typename X, typename... Args>
auto Function<T>::apply(Args&&... args) -> std::enable_if_t<std::is_same<X,T>::value, forward_t<X,Args...>> {
  // ...
  extract_vars(node->is_variable_input_, input_vars, args...);
  // ...
}
```

We need a variant of `ExtractVariables` for TensorList.

**Comment by albanD** [comment]
> But PyTorch in-source ATen operator miopen_rnn works?

These operators are not using custom Function to implement autograd. We use codegen for them which gives us a bit more freedom.

> We need a variant of ExtractVariables for TensorList.

There are other discussions about this and I don't think we want this.
Both on the python and c++ side, we have a strict rule on the autograd side that we do not "unpack" container structures (list, dict, vector, tuple, etc). The main reason for that is that are:
- How deep should you go? And what happens with circular references in these structures?
- How much of this structure should be reflected in the backward function?
- In c++ it makes the type signature for the backward function impossible to handle.
- It would make the apply() code a nightmare to match the structure and properly root every value in these structures

The simple work-around here is to make your forward function takes varargs. You can then pass an unpacked version of your vector and return that variable number of results for the backward.

**Comment by arthuryuan1987** [comment]
> The simple work-around here is to make your forward function takes varargs. You can then pass an unpacked version of your vector and return that variable number of results for the backward.

It is not a static varargs. It is dynamic varargs. Statically, template could not parse all cases. It should not work.

I mean we do not need to unpack the TensorList, and just count them. And count number of tensors could align between forward and backward. That is enough.

**Comment by xiaolil1** [comment]
I created a MR to enable extract variable list as multiple variables for custom autograd.
[https://github.com/pytorch/pytorch/pull/60926](url)

**Comment by xiaolil1** [comment]
It shows the variable list extract can solve our issue for using TensorList as parameters in some operators like LSTM and GRU, when use custom autograd. Otherwise we will meet "incorrect number of gradients" because VariableList was not treated as variable.

**Comment by albanD** [comment]
I am sure you could make few more code changes to get this use case to work. My point is not that you cannot make it work but that I don't think this is a good change for this API. It makes it significantly more complex, confusing and makes the code harder to read.
Custom functions are hard enough to use properly already.

If you really need this feature, maybe the best thing to do is to do extend the custom Function in your own code and add this specific logic there?

**Comment by arthuryuan1987** [comment]
Ok, let me clarify our case. The number of LSTM's weights is decided at runtime. We don't know the exact number statistically. So,
- there is a TensorList in prototype of aten::listm/aten::miopen_rnn, or we can keep them flatten in prototype.
- it does not work to use template.
```
template<class T>
template<typename X, typename... Args>
auto Function<T>::apply(Args&&... args) -> std::enable_if_t<std::is_same<X,T>::value, forward_t<X,Args...>> {
  ...
  extract_vars(node->is_variable_input_, input_vars, args...);
  ...
}
```
We don't know `typename... Args` statistically. We also have some try, like concatenating all tensors in TensorList into one before calling custom function,
- but, there will be performance issue, additional copies here

So have you other suggestion? Thanks.

**Comment by gujinghui** [comment]
Hi @albanD,

We are going to add LSTM for out-of-tree device.
Could you give more suggestion on this case?

Thanks.

**Comment by albanD** [comment]
Hi,

I think you have a couple solutions here. The one I would recommend as mentioned above is to extend the custom Function in your own repo and add this new feature to it directly. That way, you can update the Function::apply to do the unpacking/repacking and compute the number of inputs the way you want.

A second option that I would not recommend is to do the autograd by hand just like our codegen does it. But this will be more brittle and could lead to silent bugs.

**Comment by gujinghui** [comment]
@albanD  @ezyang 
Thanks for your suggestions.

We understood the risk of varargs in custom Function, and will follow your suggestion.

Thanks.

**Comment by albanD** [comment]
Closing this for now.
Don't hesitate to re-open if you need any help with this updated implementation!
